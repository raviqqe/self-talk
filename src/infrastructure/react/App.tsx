import { PulseLoader } from "react-spinners";
import { useAsync } from "react-use";
import styled from "styled-components";
import { Home, IProps as IHomeProps } from "./Home";
import { IProps as ILandingProps, Landing } from "./Landing";
import { white } from "./style/colors";

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
`;

export interface IProps extends IHomeProps, ILandingProps {
  initialize: () => Promise<void>;
  signedIn: boolean | null;
}

export const App = ({
  documents,
  initialize,
  repositoryURL,
  signedIn,
  signIn,
  signOut,
  ...props
}: IProps): JSX.Element => {
  useAsync(initialize, []);

  return signedIn === null ? (
    <LoaderContainer>
      <PulseLoader color={white} />
    </LoaderContainer>
  ) : signedIn ? (
    <Home {...props} documents={documents} signOut={signOut} />
  ) : (
    <Landing repositoryURL={repositoryURL} signIn={signIn} />
  );
};
