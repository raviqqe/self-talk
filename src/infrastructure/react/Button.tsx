import { defaultImport } from "default-import";
import defaultStyled from "styled-components";
import { lightGrey, red, white } from "./style/colors.js";
import { boxShadow } from "./style/index.js";

const styled = defaultImport(defaultStyled);

export const Button = styled.button<{ secondary?: boolean }>`
  ${boxShadow};
  background: ${({ secondary }) => (secondary ? lightGrey : red)};
  display: flex;
  justify-content: center;
  align-items: center;
  font-family: inherit;
  color: ${white};
  border: none;
  cursor: pointer;
  flex-shrink: 0;
`;
