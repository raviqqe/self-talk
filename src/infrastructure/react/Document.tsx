import { MdEdit } from "react-icons/md";
import React, { useState } from "react";
import styled from "styled-components";
import { IconButton } from "./IconButton";
import { Markdown } from "./Markdown";
import {
  UpdateDocument,
  IProps as IUpdateDocumentProps,
} from "./UpdateDocument";
import { boxShadow } from "./style";
import { white } from "./style/colors";

const Container = styled.div`
  ${boxShadow};
  background: ${white};
  padding: 1em;
  padding-right: 2.1em;
  border-radius: 0.5em;
  position: relative;
`;

const ButtonContainer = styled.div`
  display: flex;
  position: absolute;
  top: 0.4em;
  right: 0.4em;
`;

interface IProps extends IUpdateDocumentProps {}

export const Document = ({
  document,
  insertFiles,
  updateDocument,
  ...restProps
}: IProps) => {
  const [editing, setEditing] = useState(false);

  return editing ? (
    <UpdateDocument
      document={document}
      insertFiles={insertFiles}
      updateDocument={async (document) => {
        setEditing(false);
        await updateDocument(document);
      }}
      {...restProps}
    />
  ) : (
    <Container {...restProps}>
      <Markdown>{document.text}</Markdown>
      <ButtonContainer>
        <IconButton aria-label="Edit" onClick={() => setEditing(true)}>
          <MdEdit />
        </IconButton>
      </ButtonContainer>
    </Container>
  );
};
