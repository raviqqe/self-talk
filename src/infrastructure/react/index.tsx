import { createRoot, Root } from "react-dom/client";
import { ApplicationInitializer } from "../../application/application-initializer";
import { DocumentCreator } from "../../application/document-creator";
import { DocumentLister } from "../../application/document-lister";
import { DocumentUpdater } from "../../application/document-updater";
import { SignInManager } from "../../application/sign-in-manager";
import { SignOutManager } from "../../application/sign-out-manager";
import { TextFileInserter } from "../../application/text-file-inserter";
import { IDocument } from "../../domain/document";
import { IRenderer } from "../renderer";
import { App, IProps as IAppProps } from "./App";
import { GlobalStyle } from "./style";

interface IPresenter {
  setRenderer(renderer: IRenderer): void;
}

interface IProps extends Pick<IAppProps, "documents" | "signedIn"> {}

export class ReactRenderer implements IRenderer {
  private readonly root: Root;
  private props: IProps = { documents: null, signedIn: null };

  constructor(
    element: HTMLElement,
    presenters: IPresenter[],
    private readonly applicationInitializer: ApplicationInitializer,
    private readonly documentCreator: DocumentCreator,
    private readonly documentLister: DocumentLister,
    private readonly documentUpdater: DocumentUpdater,
    private readonly signInManager: SignInManager,
    private readonly signOutManager: SignOutManager,
    private readonly textFileInserter: TextFileInserter,
    private readonly repositoryURL: string
  ) {
    for (const presenter of presenters) {
      presenter.setRenderer(this);
    }

    this.root = createRoot(element);
  }

  public render(): void {
    this.renderProps({});
  }

  public renderDocuments(documents: IDocument[] | null): void {
    this.renderProps({ documents });
  }

  public renderSignedIn(signedIn: boolean): void {
    this.renderProps({ signedIn });
  }

  private renderProps(props: Partial<IProps>): void {
    this.props = { ...this.props, ...props };

    this.root.render(
      <>
        <App
          {...this.props}
          createDocument={(text: string) => this.documentCreator.create(text)}
          initialize={() => this.applicationInitializer.initialize()}
          insertFiles={(
            text: string,
            position: number,
            files: File[]
          ): Promise<string> =>
            this.textFileInserter.insert(text, position, files)
          }
          listDocuments={() => this.documentLister.list()}
          listMoreDocuments={() => this.documentLister.listMore()}
          repositoryURL={this.repositoryURL}
          signIn={() => this.signInManager.signIn()}
          signOut={() => this.signOutManager.signOut()}
          updateDocument={(document: IDocument) =>
            this.documentUpdater.update(document)
          }
        />
        <GlobalStyle />
      </>
    );
  }
}
