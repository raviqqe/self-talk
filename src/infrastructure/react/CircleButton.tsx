import styled from "styled-components";
import { Button } from "./Button.js";

export const CircleButton = styled(Button)`
  font-size: 1.5em;
  padding: 0;
  height: 2em;
  width: 2em;
  border-radius: 1em;
`;
