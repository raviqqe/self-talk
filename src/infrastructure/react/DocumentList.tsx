import { defaultImport } from "default-import";
import { PulseLoader } from "react-spinners";
import { useAsync } from "react-use";
import defaultStyled from "styled-components";
import type * as domain from "../../domain.js";
import { Document } from "./Document.js";
import { white } from "./style/colors.js";
import { type InsertFilesFunction } from "./utilities.js";
import defaultUseInfiniteScroll from "react-infinite-scroll-hook";

const useInfiniteScroll = defaultImport(defaultUseInfiniteScroll);
const styled = defaultImport(defaultStyled);

const Container = styled.div`
  display: flex;
  flex-direction: column-reverse;
  overflow: auto;
  padding: 1em 0.5em;
`;

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 1rem;
`;

const StyledDocument = styled(Document)`
  margin: 0.5em;
`;

export interface Props {
  documents: domain.Document[] | null;
  insertFiles: InsertFilesFunction;
  listDocuments: () => Promise<void>;
  listMoreDocuments: () => Promise<void>;
  updateDocument: (document: domain.Document) => Promise<void>;
}

export const DocumentList = ({
  documents,
  insertFiles,
  listDocuments,
  listMoreDocuments,
  updateDocument,
}: Props): JSX.Element => {
  useAsync(listDocuments, []);

  const [ref] = useInfiniteScroll({
    loading: false,
    hasNextPage: true,
    onLoadMore: listMoreDocuments,
  });

  return (
    <Container>
      {documents?.map((document) => (
        <StyledDocument
          document={document}
          insertFiles={insertFiles}
          key={document.id}
          updateDocument={updateDocument}
        />
      ))}
      <LoaderContainer ref={ref}>
        <PulseLoader color={white} />
      </LoaderContainer>
    </Container>
  );
};
