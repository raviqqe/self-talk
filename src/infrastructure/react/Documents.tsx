import { PulseLoader } from "react-spinners";
import InfiniteScroller from "react-infinite-scroller";
import { useAsync } from "react-use";
import React from "react";
import styled from "styled-components";
import { IDocument } from "../../domain/document";
import { InsertFilesFunction } from "./utilities";
import { Document } from "./Document";
import { white } from "./style/colors";

const Container = styled.div`
  display: flex;
  flex-direction: column-reverse;
  overflow: auto;
`;

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

const StyledInfiniteScroller = styled(InfiniteScroller)`
  display: flex;
  flex-direction: column-reverse;
  padding: 1em 0.5em;
`;

const StyledDocument = styled(Document)`
  margin: 0.5em;
`;

export interface IProps {
  documents: IDocument[] | null;
  insertFiles: InsertFilesFunction;
  listDocuments: () => Promise<void>;
  listMoreDocuments: () => Promise<void>;
  updateDocument: (document: IDocument) => Promise<void>;
}

export const Documents = ({
  documents,
  insertFiles,
  listDocuments,
  listMoreDocuments,
  updateDocument,
}: IProps) => {
  useAsync(listDocuments, []);

  return documents ? (
    <Container>
      <StyledInfiniteScroller
        hasMore={true}
        isReverse={true}
        loadMore={listMoreDocuments}
        threshold={512}
        useWindow={false}
      >
        {documents.map((document: IDocument) => (
          <StyledDocument
            document={document}
            insertFiles={insertFiles}
            key={document.id}
            updateDocument={updateDocument}
          />
        ))}
      </StyledInfiniteScroller>
    </Container>
  ) : (
    <LoaderContainer>
      <PulseLoader color={white} />
    </LoaderContainer>
  );
};
