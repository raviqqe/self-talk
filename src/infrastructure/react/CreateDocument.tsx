import { useState } from "react";
import { MdAdd } from "react-icons/md";
import styled from "styled-components";
import { CircleButton } from "./CircleButton";
import { MarkdownTextArea } from "./MarkdownTextArea";
import { InsertFilesFunction } from "./utilities";

const Container = styled.div`
  display: flex;
  align-items: center;
  padding: 1em;
`;

const StyledMarkdownTextArea = styled(MarkdownTextArea)`
  margin-right: -1em;
  max-height: 80vh;
`;

interface IProps {
  createDocument: (text: string) => Promise<void>;
  insertFiles: InsertFilesFunction;
}

export const CreateDocument = ({
  createDocument,
  insertFiles,
}: IProps): JSX.Element => {
  const [text, setText] = useState("");
  const onSubmit = async (): Promise<void> => {
    setText("");
    await createDocument(text);
  };

  return (
    <Container>
      <StyledMarkdownTextArea
        insertFiles={insertFiles}
        onSubmit={onSubmit}
        setText={setText}
        text={text}
      />
      <CircleButton aria-label="Create" onClick={onSubmit}>
        <MdAdd />
      </CircleButton>
    </Container>
  );
};
