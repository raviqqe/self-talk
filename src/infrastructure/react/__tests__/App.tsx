import { render } from "@testing-library/react";
import React from "react";
import { App, IProps } from "../App";

const props: IProps = {
  createDocument: async () => {},
  documents: null,
  initialize: async () => {},
  insertFiles: async () => "",
  listDocuments: async () => {},
  listMoreDocuments: async () => {},
  repositoryURL: "",
  signIn: async () => {},
  signOut: async () => {},
  signedIn: null,
  updateDocument: async () => {},
};

it("renders before a user signs in", () => {
  expect(
    render(<App {...props} signedIn={null} />).container
  ).toMatchSnapshot();
});

it("renders after a user signs in", () => {
  expect(
    render(<App {...props} signedIn={true} />).container
  ).toMatchSnapshot();
});

it("renders after a user signs out", () => {
  expect(
    render(<App {...props} signedIn={false} />).container
  ).toMatchSnapshot();
});
