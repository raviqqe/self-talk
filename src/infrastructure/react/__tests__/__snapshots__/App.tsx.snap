// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders after a user signs in 1`] = `
<div>
  <div
    class="sc-fzpjYC fVAfyo"
  >
    <div
      class="sc-fzoyAV gTIEqv"
    >
      <div
        class="css-0"
      >
        <div
          class="css-b0t2tw"
        />
        <div
          class="css-fz02ge"
        />
        <div
          class="css-1xlchjq"
        />
      </div>
    </div>
    <div
      class="sc-fznxsB loNfYY"
    >
      <div
        class="sc-fzoLsD bJfDCP"
      >
        <textarea
          async=""
          class="sc-AxmLO jUPA-dU sc-fzpans eCQXVP"
          placeholder="Write in Markdown ..."
          rows="1"
        />
        <button
          aria-label="Create"
          class="sc-AxjAm sc-Axmtr jmpsoc"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </button>
      </div>
      <div
        class="sc-fznJRM dWQTMH"
      />
    </div>
    <div
      class="sc-fznWqX LbrhL"
    >
      <button
        class="sc-AxjAm sc-Axmtr hidflr"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`renders after a user signs out 1`] = `
<div>
  <div
    class="sc-AxiKw lgKAvH"
  >
    <div
      class="sc-AxhCb eiDjsk"
    >
      <span
        class="sc-AxhUy djJDyD"
      >
        Self
      </span>
      <span
        class="sc-AxgMl goKvzw"
      >
        Talk
      </span>
    </div>
    <button
      class="sc-AxjAm sc-AxirZ hIdFsg"
    >
      Sign in
    </button>
    <a
      class="sc-AxheI fSlMcm"
      href=""
      target="_blank"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"
          fill-rule="evenodd"
        />
      </svg>
    </a>
  </div>
</div>
`;

exports[`renders before a user signs in 1`] = `
<div>
  <div
    class="sc-fzoiQi kYjXrO"
  >
    <div
      class="css-0"
    >
      <div
        class="css-b0t2tw"
      />
      <div
        class="css-fz02ge"
      />
      <div
        class="css-1xlchjq"
      />
    </div>
  </div>
</div>
`;
